File: WFM-Planning-Suite/web/src/core/cleansing/strategies.ts
type ImputeFn = (values: number[], anomalies: number[]) => [number[], Array<{ index: number; old: number; new: number; method: string }>]

function median(nums: number[]): number {
  const s = [...nums].sort((a, b) => a - b)
  const m = Math.floor(s.length / 2)
  return s.length % 2 ? s[m] : (s[m - 1] + s[m]) / 2
}

function detectSpikes(values: number[], z = 3): number[] {
  if (values.length < 3) return []
  const mean = values.reduce((a, b) => a + b, 0) / values.length
  const variance = values.reduce((a, v) => a + (v - mean) ** 2, 0) / values.length
  const stdev = Math.sqrt(variance)
  if (stdev === 0) return []
  return values.map((v, i) => (Math.abs(v - mean) > z * stdev ? i : -1)).filter((i) => i >= 0)
}

const imputeMedian: ImputeFn = (values, anomalies) => {
  const result = [...values]
  const changes: Array<{ index: number; old: number; new: number; method: string }> = []
  const clean = values.filter((_, i) => !anomalies.includes(i))
  const fill = clean.length ? median(clean) : 0
  for (const idx of anomalies) {
    changes.push({ index: idx, old: result[idx], new: fill, method: 'median' })
    result[idx] = fill
  }
  return [result, changes]
}

const imputeLinear: ImputeFn = (values, anomalies) => {
  const result = [...values]
  const changes: Array<{ index: number; old: number; new: number; method: string }> = []
  const set = new Set(anomalies)
  for (const idx of [...anomalies].sort((a, b) => a - b)) {
    let prev: number | null = null, next: number | null = null
    for (let j = idx - 1; j >= 0; j--) if (!set.has(j)) { prev = result[j]; break }
    for (let j = idx + 1; j < result.length; j++) if (!set.has(j)) { next = result[j]; break }
    const nv = prev !== null && next !== null ? (prev + next) / 2 : prev ?? next ?? 0
    changes.push({ index: idx, old: result[idx], new: nv, method: 'linear_interpolation' })
    result[idx] = nv
  }
  return [result, changes]
}

const imputeRolling: ImputeFn = (values, anomalies) => {
  const result = [...values]
  const changes: Array<{ index: number; old: number; new: number; method: string }> = []
  const set = new Set(anomalies)
  for (const idx of anomalies) {
    const neighbors = values.filter((_, j) => Math.abs(j - idx) <= 5 && !set.has(j))
    const nv = neighbors.length ? neighbors.reduce((a, b) => a + b, 0) / neighbors.length : 0
    changes.push({ index: idx, old: result[idx], new: nv, method: 'rolling_mean' })
    result[idx] = nv
  }
  return [result, changes]
}

export const IMPUTATION_METHODS: Record<string, ImputeFn> = {
  median: imputeMedian,
  linear_interpolation: imputeLinear,
  rolling_mean: imputeRolling,
}

export function applyCleansing(values: number[], method: string, zThreshold = 3) {
  const anomalies = [...new Set(detectSpikes(values, zThreshold))]
  const fn = IMPUTATION_METHODS[method]
  if (!fn) throw new Error(`Unknown method: ${method}`)
  const [cleaned, changes] = fn(values, anomalies)
  return { cleaned, changes, anomalies }
}

File: WFM-Planning-Suite/src/core/cleansing/strategies.py
"""Anomaly detection and genuinely distinct imputation strategies."""

from __future__ import annotations

import statistics
from typing import Callable


def detect_spikes(values: list[float], z_threshold: float = 3.0) -> list[int]:
    if len(values) < 3:
        return []
    mean = statistics.mean(values)
    stdev = statistics.stdev(values)
    if stdev == 0:
        return []
    return [i for i, v in enumerate(values) if abs(v - mean) > z_threshold * stdev]


def detect_zeros_runs(values: list[float], min_run: int = 3) -> list[int]:
    indices: list[int] = []
    run_start = None
    for i, v in enumerate(values):
        if v == 0:
            if run_start is None:
                run_start = i
        else:
            if run_start is not None and i - run_start >= min_run:
                indices.extend(range(run_start, i))
            run_start = None
    if run_start is not None and len(values) - run_start >= min_run:
        indices.extend(range(run_start, len(values)))
    return indices


def impute_median(values: list[float], anomaly_indices: list[int]) -> tuple[list[float], list[dict]]:
    result = list(values)
    changes: list[dict] = []
    clean = [v for i, v in enumerate(values) if i not in anomaly_indices and v is not None]
    fill = statistics.median(clean) if clean else 0.0
    for idx in anomaly_indices:
        old = result[idx]
        result[idx] = fill
        changes.append({"index": idx, "old": old, "new": fill, "method": "median"})
    return result, changes


def impute_linear_interpolation(values: list[float], anomaly_indices: list[int]) -> tuple[list[float], list[dict]]:
    result = list(values)
    changes: list[dict] = []
    anomaly_set = set(anomaly_indices)
    for idx in sorted(anomaly_indices):
        old = result[idx]
        prev_val = None
        next_val = None
        for j in range(idx - 1, -1, -1):
            if j not in anomaly_set:
                prev_val = result[j]
                break
        for j in range(idx + 1, len(result)):
            if j not in anomaly_set:
                next_val = result[j]
                break
        if prev_val is not None and next_val is not None:
            new = (prev_val + next_val) / 2.0
        elif prev_val is not None:
            new = prev_val
        elif next_val is not None:
            new = next_val
        else:
            new = 0.0
        result[idx] = new
        changes.append({"index": idx, "old": old, "new": new, "method": "linear_interpolation"})
    return result, changes


def impute_rolling_mean(values: list[float], anomaly_indices: list[int], window: int = 5) -> tuple[list[float], list[dict]]:
    result = list(values)
    changes: list[dict] = []
    anomaly_set = set(anomaly_indices)
    for idx in anomaly_indices:
        old = result[idx]
        neighbors = [
            result[j]
            for j in range(max(0, idx - window), min(len(result), idx + window + 1))
            if j not in anomaly_set
        ]
        new = statistics.mean(neighbors) if neighbors else 0.0
        result[idx] = new
        changes.append({"index": idx, "old": old, "new": new, "method": "rolling_mean"})
    return result, changes


IMPUTATION_METHODS: dict[str, Callable[[list[float], list[int]], tuple[list[float], list[dict]]]] = {
    "median": impute_median,
    "linear_interpolation": impute_linear_interpolation,
    "rolling_mean": impute_rolling_mean,
}


def apply_cleansing(
    values: list[float],
    method: str,
    z_threshold: float = 3.0,
) -> tuple[list[float], list[dict], list[int]]:
    anomalies = list(set(detect_spikes(values, z_threshold) + detect_zeros_runs(values)))
    if method not in IMPUTATION_METHODS:
        raise ValueError(f"Unknown cleansing method: {method}")
    cleaned, changes = IMPUTATION_METHODS[method](values, anomalies)
    return cleaned, changes, anomalies
